src/gen_rules.ml

```diff
@@ -315,14 +351,20 @@ let gen ~contexts
         Fiber.Ivar.read (Hashtbl.find_exn sctxs h.name)
         >>| Option.some
     in
+
     let stanzas () =
       let+ stanzas = Dune_load.Dune_files.eval ~context dune_files in
       match only_packages with
       | None -> stanzas
       | Some pkgs ->
+        let relevant_lib_names =
+          stanzas
+          |> List.concat_map ~f:(fun (dir_conf : Dune_load.Dune_file.t) -> dir_conf.stanzas)
+          |> build_name_set pkgs
```

Comment:
So I added a `Dune_load.Dune_file.fold_stanzas` that acts like deep_fold

Reply:
Looks good, but we can go one step further: instead of creating a list, filtering it and then turning it to a set, we can fold a set right from the start.

I'm expecting that when we start using dune on very large workspaces these things will matter.

Reply:
Oh yes of course I did not see that